[APIv3] add documentation for attachments

NobodysNightmare opened the pull request:
## OpenProject work package

https://community.openproject.org/work_packages/19988

NobodysNightmare committed:
specify the attachments resource

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://nobodysnightmare-feature-specify-attachments.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://nobodysnightmare-feature-specify-attachments.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

NobodysNightmare committed:
specify attachment listing per work package

NobodysNightmare committed:
rename "Properties" heading

Properties has become a MSON keyword and therefore produces compile warnings when used.

We could either enclose it in backticks or rename it.
I went for renaming, because that makes the distinction between "linked" and "non-linked" properties more apparent

NobodysNightmare committed:
fix spelling

Azure7111 commented on doc/apiv3-documentation.apib:
```diff
@@ -552,17 +572,24 @@ Updates an activity's comment and, on success, returns the updated activity.
+|:----------------:| --------------------------------------------------- | ------------- | ----------- | -------------------- |
+| self             | This attachment                                     | Attachment    | not null    | READ                 |
+| container        | The object (e.g. WorkPackage) housing the attachment| Anything      | not null    | READ                 |
+| author           | The user that uploaded the attachment               | User          | not null    | READ                 |
```
This should probably read "..user who..."

Azure7111 commented on doc/apiv3-documentation.apib:
```diff
@@ -552,17 +572,24 @@ Updates an activity's comment and, on success, returns the updated activity.
+| self             | This attachment                                     | Attachment    | not null    | READ                 |
+| container        | The object (e.g. WorkPackage) housing the attachment| Anything      | not null    | READ                 |
+| author           | The user that uploaded the attachment               | User          | not null    | READ                 |
+| downloadLocation | Direct download link to the attachment              | -             | not null    | READ                 |
```
No Type? I'd expected something like URL or String.

NobodysNightmare replied:
String would not be correct, because I am classifying it as a link, thus the type indicates of whcih type the _linked_ resource is. But in this case the link does not point into the API at all.

An alternative would be to not provide this as a link, but instead as a local property, which is a `String`.

At the moment I believe that it is cleaner toprovide a download URL as `_link`, even if this means the API documentation lacks a type.

Azure7111 replied:
Didn't know that!

This sounds reasonable, let future us worry about the type issue!

Azure7111 commented on doc/apiv3-documentation.apib:
```diff
@@ -552,17 +572,24 @@ Updates an activity's comment and, on success, returns the updated activity.
 
 # Group Attachments
```
No flavour text for attachments? I mean, yes this is pretty basic and obvious stuff, but for the sake of consistency I'd expected some introductory text.

Azure7111 commented on doc/apiv3-documentation.apib:
```diff
@@ -573,26 +600,28 @@ Updates an activity's comment and, on success, returns the updated activity.
-                        "href": "/api/v3/users/1",
-                        "title": "John Sheppard - admin"
+                        "href": "/api/v3/users/1"
                     }
```
I think a downloadlocation link should be included here.

Azure7111 commented on doc/apiv3-documentation.apib:
```diff
@@ -605,6 +634,176 @@ Updates an activity's comment and, on success, returns the updated activity.
+
+## Delete attachment [DELETE]
+
+Permanently deletes the specified user account.
```
The words "user account" should probably be replaced by the word "attachment"

Azure7111 commented on doc/apiv3-documentation.apib:
```diff
@@ -605,6 +634,176 @@ Updates an activity's comment and, on success, returns the updated activity.
+                }
+            }
+
+## List attachments of a work package [GET]
```
As the linked work package is about CRUD for attachments you might want to add a section about POSTing new attachments to a work package ;)

Azure7111 commented on doc/apiv3-documentation.apib:
```diff
@@ -605,6 +634,176 @@ Updates an activity's comment and, on success, returns the updated activity.
+                                },
+                                "author": {
+                                    "href": "/api/v3/users/1"
+                                }
```
You might want to add a downloadlocation link as well.

Azure7111 commented on doc/apiv3-documentation.apib:
```diff
@@ -605,6 +634,176 @@ Updates an activity's comment and, on success, returns the updated activity.
+                                },
+                                "author": {
+                                    "href": "/api/v3/users/1"
+                                }
```
Missing downloadlocation link

Azure7111 commented on doc/apiv3-documentation.apib:
```diff
@@ -93,7 +93,7 @@ For the future we plan to add authentication modes based on OAuth2 and OpenID Co
 Links to other resources in the API are represented uniformly by so called link objects.
 
-## Properties
+## Local Properties
```
If you want to be thorough: There are a few places like this we you either have a : appended and then some others without. You may want to consider making sure that its consistent across the entire document. This also applies to captions like Linked Properties.

NobodysNightmare commented:
closing in favor of a new PR, from a new branch
